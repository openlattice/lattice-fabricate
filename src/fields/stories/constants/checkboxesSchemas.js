export const schema = {
  type: 'object',
  title: 'CheckboxArrayField',
  properties: {
    checkboxes: {
      title: 'Options',
      type: 'array',
      items: {
        type: 'string',
        enum: ['Option 1', 'Option 2', 'Option 3', 'Other'],
      },
      minItems: 1,
      uniqueItems: true
    },
  },
};

export const uiSchema = {
  classNames: 'column-span-12',
  checkboxes: {
    classNames: 'column-span-12',
    'ui:field': 'CheckboxArrayField',
    'ui:options': {
      withOther: true
    }
  }
};
