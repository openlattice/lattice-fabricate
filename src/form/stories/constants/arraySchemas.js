export const schema = {
  type: 'object',
  title: 'Task List',
  properties: {
    tasks: {
      title: 'Tasks',
      type: 'array',
      items: {
        type: 'object',
        properties: {
          taskName: {
            type: 'string',
            title: 'Task Name',
            default: ''
          },
          taskDescription: {
            type: 'string',
            title: 'Task Description',
            default: ''
          }
        }
      },
      default: [{
        taskName: '',
        taskDescription: ''
      }]
    },
  }
};

export const uiSchema = {
  tasks: {
    classNames: 'column-span-12',
    items: {
      classNames: 'grid-container',
      taskName: {
        classNames: 'column-span-12'
      },
      taskDescription: {
        classNames: 'column-span-12',
        'ui:widget': 'textarea'
      }
    }
  }
};
