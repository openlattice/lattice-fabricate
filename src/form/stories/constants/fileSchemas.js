// @flow

const schema = {
  title: 'Files',
  type: 'object',
  properties: {
    file: {
      type: 'string',
      format: 'data-url',
      title: 'Single file'
    },
    files: {
      type: 'array',
      title: 'Multiple files',
      items: {
        type: 'string',
        format: 'data-url'
      }
    },
    filesAccept: {
      type: 'string',
      format: 'data-url',
      title: 'Single File with Accept attribute'
    }
  }
};

const uiSchema = {
  file: {
    classNames: 'column-span-12',
  },
  files: {
    classNames: 'column-span-12',
  },
  filesAccept: {
    classNames: 'column-span-12',
    'ui:options': {
      accept: '.pdf'
    }
  }
};

export {
  schema,
  uiSchema
};
