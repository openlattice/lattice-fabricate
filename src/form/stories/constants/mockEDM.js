// @flow
import type { FQN } from 'lattice';

import {
  ASSOCIATION_ENTITY_SET_NAMES,
  ENTITY_SET_NAMES,
  PROPERTY_TYPE_FQNS
} from './mockFQNs';

let value = 0;
const UUID_TEMPLATE = '00000000-0000-0000-0000-000000000000';
function mockUUID() {
  value += 1;
  const valueStr = value.toString();
  const fakeUUID = UUID_TEMPLATE.slice(0, UUID_TEMPLATE.length - valueStr.length) + valueStr;
  return fakeUUID;
}

const allEntitySetNames :string[] = (Object.values(ENTITY_SET_NAMES)
  .concat(Object.values(ASSOCIATION_ENTITY_SET_NAMES)) :any);

const entitySetIds = allEntitySetNames.reduce((acc, entitySetNames) => {
  acc[entitySetNames] = mockUUID();
  return acc;
}, {});

const allPropertyTypeFQNs :FQN[] = (Object.values(PROPERTY_TYPE_FQNS) :any);
const propertyTypeIds = allPropertyTypeFQNs.reduce((acc, propertyTypeFQN) => {
  // $FlowFixMe
  acc[propertyTypeFQN] = mockUUID();
  return acc;
}, {});

export {
  entitySetIds,
  propertyTypeIds
};
