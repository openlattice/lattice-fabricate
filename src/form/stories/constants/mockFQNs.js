import { Models } from 'lattice';

const { FullyQualifiedName } = Models;

const PROPERTY_TYPE_FQNS = {
  COMPLETED_DT_FQN: new FullyQualifiedName('mock.completeddatetime'),
  DESCRIPTION_FQN: new FullyQualifiedName('mock.description'),
  IMAGE_DATA_FQN: new FullyQualifiedName('mock.imagedata'),
  INDEX_FQN: new FullyQualifiedName('mock.index'),
  NAME_FQN: new FullyQualifiedName('mock.name'),
};

const ENTITY_SET_NAMES = {
  IMAGE_ESN: 'mock_image',
  PERSON_ESN: 'mock_person',
  TASK_ESN: 'mock_task',
  TASK_LIST_ESN: 'mock_tasklist',
};

const ASSOCIATION_ENTITY_SET_NAMES = {
  INCLUDES_ESN: 'mock_includes'
};

export {
  ASSOCIATION_ENTITY_SET_NAMES,
  ENTITY_SET_NAMES,
  PROPERTY_TYPE_FQNS,
};
