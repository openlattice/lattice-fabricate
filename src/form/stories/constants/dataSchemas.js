import { ENTITY_SET_NAMES, PROPERTY_TYPE_FQNS } from './mockFQNs';

import { getEntityAddressKey, getPageSectionKey } from '../../../utils/DataProcessingUtils';

const {
  TASK_ESN,
} = ENTITY_SET_NAMES;

const {
  NAME_FQN,
  DESCRIPTION_FQN,
  INDEX_FQN,
  TYPE_FQN,
} = PROPERTY_TYPE_FQNS;

export const schema = {
  definitions: {
    taskItems: {
      type: 'object',
      properties: {
        [getEntityAddressKey(-1, TASK_ESN, NAME_FQN)]: {
          type: 'number',
          title: 'Task Name',
        },
        [getEntityAddressKey(-1, TASK_ESN, DESCRIPTION_FQN)]: {
          type: 'string',
          title: 'Task Description',
        },
        [getEntityAddressKey(-1, TASK_ESN, TYPE_FQN)]: {
          type: 'array',
          title: 'Notifications',
          items: {
            type: 'string',
            enum: ['Push', 'Email']
          },
          uniqueItems: true
        },
        [getEntityAddressKey(-1, TASK_ESN, INDEX_FQN)]: {
          type: 'number',
          title: 'Task #',
        }
      },
      required: [getEntityAddressKey(-1, TASK_ESN, NAME_FQN)],
    }
  },
  type: 'object',
  title: 'Arrays',
  properties: {
    [getPageSectionKey(1, 1)]: {
      type: 'array',
      title: 'Tasks',
      items: {
        $ref: '#/definitions/taskItems'
      },
      default: [
        {
          [getEntityAddressKey(-1, TASK_ESN, INDEX_FQN)]: 0
        },
      ]
    }
  }
};

export const uiSchema = {
  [getPageSectionKey(1, 1)]: {
    classNames: 'column-span-12',
    'ui:options': {
      addActionKey: 'addTaskItem',
      addButtonText: '+ Add Task'
    },
    items: {
      classNames: 'grid-container',
      [getEntityAddressKey(-1, TASK_ESN, NAME_FQN)]: {
        classNames: 'column-span-12'
      },
      [getEntityAddressKey(-1, TASK_ESN, DESCRIPTION_FQN)]: {
        classNames: 'column-span-12',
        'ui:widget': 'textarea'
      },
      [getEntityAddressKey(-1, TASK_ESN, INDEX_FQN)]: {
        classNames: 'column-span-12',
        'ui:widget': 'hidden'
      },
      [getEntityAddressKey(-1, TASK_ESN, TYPE_FQN)]: {
        classNames: 'column-span-12',
        'ui:widget': 'checkboxes',
        'ui:options': {
          withNone: true,
          row: true,
          mode: 'button'
        }
      },
      'ui:options': {
        editable: true
      }
    }
  }
};
