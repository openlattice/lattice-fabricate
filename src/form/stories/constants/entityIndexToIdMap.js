import { Map } from 'immutable';
import { v4 as uuid } from 'uuid';

import { ENTITY_SET_NAMES } from './mockFQNs';

const { TASK_ESN } = ENTITY_SET_NAMES;

const entityIndexToIdMap = {
  [TASK_ESN]: Map([
    // there are two unique entities in a list where the EAK index of the list is -1
    [-1, [uuid(), uuid()]],
  ])
};


export default entityIndexToIdMap;
