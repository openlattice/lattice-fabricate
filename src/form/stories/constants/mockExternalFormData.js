// @flow

import { getEntityAddressKey, getPageSectionKey } from '../../../utils/DataProcessingUtils';
import { ENTITY_SET_NAMES, PROPERTY_TYPE_FQNS } from './mockFQNs';

const {
  TASK_ESN,
} = ENTITY_SET_NAMES;

const {
  NAME_FQN,
  DESCRIPTION_FQN,
  INDEX_FQN
} = PROPERTY_TYPE_FQNS;

const mockExternalFormData = {
  [getPageSectionKey(1, 1)]: [
    {
      [getEntityAddressKey(-1, TASK_ESN, NAME_FQN)]: 123,
      [getEntityAddressKey(-1, TASK_ESN, DESCRIPTION_FQN)]: undefined,
      [getEntityAddressKey(-1, TASK_ESN, INDEX_FQN)]: 0
    },
    {
      [getEntityAddressKey(-1, TASK_ESN, NAME_FQN)]: 456,
      [getEntityAddressKey(-1, TASK_ESN, DESCRIPTION_FQN)]: undefined,
      [getEntityAddressKey(-1, TASK_ESN, INDEX_FQN)]: 1
    }
  ]
};

export default mockExternalFormData;
