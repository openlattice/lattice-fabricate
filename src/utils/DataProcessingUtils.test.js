/*
 * @flow
 */

import { List, Map } from 'immutable';
import { Models } from 'lattice';

import {
  VALUE_MAPPERS,
  getEntityAddressKey,
  parseEntityAddressKey,
  processEntityData,
  processEntityValue
} from './DataProcessingUtils';
import {
  INVALID_PARAMS,
  INVALID_PARAMS_FOR_REQUIRED_INDEX_OR_SS,
  INVALID_PARAMS_SS,
} from './testing/Invalid';

import mockExternalFormData from '../form/stories/constants/mockExternalFormData';
import { entitySetIds, propertyTypeIds } from '../form/stories/constants/mockEDM';

const { FQN } = Models;

const MOCK_EKID = '9b93bc80-79c3-44c8-807c-ada1a8d6484f';
const MOCK_ESN = 'MockEntitySetName';
const MOCK_FQN = FQN.of('ol.mock');

describe('DataProcessingUtils', () => {

  describe('getEntityAddressKey()', () => {

    test('should return a valid entity address key given an index', () => {
      for (let mockIndex = -3; mockIndex < 4; mockIndex += 1) {
        const eak = getEntityAddressKey(mockIndex, MOCK_ESN, MOCK_FQN);
        // $FlowFixMe
        expect(eak).toEqual(`${mockIndex}__@@__${MOCK_ESN}__@@__${MOCK_FQN}`);
      }
    });

    test('should return a valid entity address key given an entity key id', () => {
      const eak = getEntityAddressKey(MOCK_EKID, MOCK_ESN, MOCK_FQN);
      // $FlowFixMe
      expect(eak).toEqual(`${MOCK_EKID}__@@__${MOCK_ESN}__@@__${MOCK_FQN}`);
    });

    test('should throw given an invalid index or entity key id', () => {
      INVALID_PARAMS_FOR_REQUIRED_INDEX_OR_SS.forEach((invalidParam) => {
        expect(() => {
          getEntityAddressKey(invalidParam, MOCK_ESN, MOCK_FQN);
        }).toThrow();
      });
    });

    test('should throw given an invalid entity set name or fqn', () => {
      INVALID_PARAMS.forEach((invalidParam) => {
        expect(() => {
          getEntityAddressKey(MOCK_EKID, invalidParam, MOCK_FQN);
        }).toThrow();
      });
    });

    test('should throw given an invalid fqn', () => {
      INVALID_PARAMS_SS.forEach((invalidParam) => {
        expect(() => {
          getEntityAddressKey(MOCK_EKID, MOCK_ESN, invalidParam);
        }).toThrow();
      });
    });

  });

  describe('parseEntityAddressKey()', () => {

    test('should return the correct components of an entity address key given an index', () => {
      for (let mockIndex = -3; mockIndex < 4; mockIndex += 1) {
        // $FlowFixMe
        const entityAddressKey = parseEntityAddressKey(`${mockIndex}__@@__${MOCK_ESN}__@@__${MOCK_FQN}`);
        expect(entityAddressKey).toEqual({
          entityIndex: mockIndex,
          entitySetName: MOCK_ESN,
          propertyTypeFQN: MOCK_FQN,
        });
      }
    });

    test('should return the correct components of an entity address key given an entity key id', () => {
      // $FlowFixMe
      const entityAddressKey = parseEntityAddressKey(`${MOCK_EKID}__@@__${MOCK_ESN}__@@__${MOCK_FQN}`);
      expect(entityAddressKey).toEqual({
        entityKeyId: MOCK_EKID,
        entitySetName: MOCK_ESN,
        propertyTypeFQN: MOCK_FQN,
      });
    });

    test('should throw given an entity address key with invalid entity key id', () => {
      INVALID_PARAMS_FOR_REQUIRED_INDEX_OR_SS.forEach((invalidParam) => {
        expect(() => {
          // $FlowFixMe
          parseEntityAddressKey(`${invalidParam}__@@__${MOCK_ESN}__@@__${MOCK_FQN}`);
        }).toThrow();
      });
    });

    test('should throw given an entity address key with invalid entity set name', () => {
      expect(() => {
        // $FlowFixMe
        parseEntityAddressKey(`${MOCK_EKID}__@@__${''}__@@__${MOCK_FQN}`);
      }).toThrow();
      expect(() => {
        // $FlowFixMe
        parseEntityAddressKey(`${MOCK_EKID}__@@__${' '}__@@__${MOCK_FQN}`);
      }).toThrow();
    });

    test('should throw given an entity address key with invalid fqn', () => {
      expect(() => {
        parseEntityAddressKey(`${MOCK_EKID}__@@__${MOCK_ESN}__@@__${''}`);
      }).toThrow();
      expect(() => {
        parseEntityAddressKey(`${MOCK_EKID}__@@__${MOCK_ESN}__@@__${' '}`);
      }).toThrow();
      expect(() => {
        parseEntityAddressKey(`${MOCK_EKID}__@@__${MOCK_ESN}__@@__${'invalid_fqn'}`);
      }).toThrow();
    });

  });

  describe('processEntityValue', () => {

    test('should return mapped value wrapped in array', () => {
      const mockMapper = jest.fn();
      const testMapper = (value) => {
        mockMapper(value);
        return value;
      };

      const mockMappers = {
        [VALUE_MAPPERS]: {
          test: testMapper
        }
      };

      const UNWRAPPED_VALUES = [
        'OpenLattice',
        ['OpenLattice'],
        List(['OpenLattice']),
        { Open: 'Lattice' },
        Map({ Open: 'Lattice' }),
        true,
        false,
        0,
        1,
      ];

      const WRAPPED_VALUES = [
        ['OpenLattice'],
        ['OpenLattice'],
        ['OpenLattice'],
        [{ Open: 'Lattice' }],
        [Map({ Open: 'Lattice' })],
        [true],
        [false],
        [0],
        [1],
      ];

      for (let i = 0; i < UNWRAPPED_VALUES.length; i += 1) {
        const value = UNWRAPPED_VALUES[i];
        const processedValue = processEntityValue('test', value, mockMappers);
        expect(mockMapper).toHaveBeenLastCalledWith(value);
        expect(processedValue).toEqual(WRAPPED_VALUES[i]);
      }
    });
  });

  describe('processEntityData', () => {
    test('should process formData', () => {
      const processedValue = processEntityData(mockExternalFormData, entitySetIds, propertyTypeIds);
      expect(processedValue).toMatchSnapshot();
    });
  });

});
