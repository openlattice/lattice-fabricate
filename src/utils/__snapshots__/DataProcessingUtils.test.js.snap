// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProcessingUtils processEntityData should process formData 1`] = `
Object {
  "00000000-0000-0000-0000-000000000003": Array [
    Object {
      "00000000-0000-0000-0000-000000000007": Array [
        "Open",
      ],
      "00000000-0000-0000-0000-000000000009": Array [
        0,
      ],
      "00000000-0000-0000-0000-000000000010": Array [
        123,
      ],
      "00000000-0000-0000-0000-000000000011": Array [
        "Push",
        "Email",
      ],
    },
    Object {
      "00000000-0000-0000-0000-000000000007": Array [
        "Lattice",
      ],
      "00000000-0000-0000-0000-000000000009": Array [
        1,
      ],
      "00000000-0000-0000-0000-000000000010": Array [
        456,
      ],
      "00000000-0000-0000-0000-000000000011": Array [
        "None",
      ],
    },
  ],
}
`;
